### ISSUE TITLE
Sympy should add an NCPoly/FreeAlgebraElement class.

### ISSUE THREAD

**Issue by zhempstead** [sympy/sympy#26114]
This issue arose from discussion on https://github.com/sympy/sympy/pull/26093, which was created to address https://github.com/sympy/sympy/issues/26041. @oscarbenjamin and I agreed that noncommutative factoring was involved enough that we should write a properly-written NCPoly class.

To kick off the discussion, here are some of @oscarbenjamin's thoughts on that PR with my responses:

> ...The interface does not need to be complete but it should have the right representation and interface on a basic level. That way people can build on it in future but the foundations will be sound.

> I am actually not sure what are common data structures or interfaces used for noncommutative polynomials. Presumably dense representations make no sense.

When implementing my not-well-grounded private _NCPoly class for the original issue, I went with a list of tuples where the first element of the tuple is the index of the generator, and the second element is the power. So if the generators are `x, y, z`, then `x**3 y x**2 z` is represented by `[(0, 3), (1, 1), (0, 2), (2, 1)]`.

Julia has an implementation which they call `FreeAssociativeAlgebra` - FreeAlgebra is the established name for a "polynomial" ring with non-commutative generators: https://en.wikipedia.org/wiki/Free_algebra. Not sure what a good name for the class would be - `FreeAlgebraPoly`? `FAPoly`? `NCPoly`? `FAElement`? etc... An element of a free algebra is not technically a polynomial so maybe best to not have `Poly` in the name?

From https://docs.juliahub.com/AbstractAlgebra/b8V2b/0.25.2/free_associative_algebra/ it *seems* like their internal representation is just a list, where each element is a generator separately. So `x**3 y x**2 z` would be represented by `[0, 0, 0, 1, 0, 0, 2]` (well, not quite because they seem to use 1-indexing, but same idea). This is obviously worse than my representation if we expect there to be high powers with many consecutive instances of the same generator (i.e. `x**20 y**10`), and better if we don't.

> I imagine that this would end up looking like SymPy's existing sparse polynomials but just with a different kinds of monomials and different polynomial operations being supported. That makes me wonder to what extent the existing sparse polynomial rings could just be adapted to handle this as an alternate case:
>
> https://github.com/sympy/sympy/blob/36bcecfeb5b359238a6047c7888555e196d43c89/sympy/polys/rings.py#L227-L245

We should definitely re-use as much code as possible. One awkward part is that these "polynomials" have both left and right division. I'm not sure if it makes sense to just arbitrarily choose e.g. left division as the default division and add a special function for right division, or to make it clear via the interface that "division" on its own is not an unambiguous instruction.

> I wonder if it makes sense to allow negative powers from the outset. I would like to know how any other algebra system handles them.

As far as I can tell, `Poly` doesn't *exactly* handle negative powers. Rather, if you have e.g. `5*x**-10` it will add `x**-1` as a generator (similarly, if `5*sin(x)` is a term, then `sin(x)` will be added as a generator). If `x` and `x**-1` are both generators, the Poly class doesn't know that the two cancel out - it doesn't know that `x + x**-1 + 2` factors into `(x + 1)*(x**-1 + 1)`, for instance.

If FreeAlgebra uses Domains properly the way Poly does, I *think* we will get this level of negative power support for free.

**Comment by zhempstead** [comment]
The `poly` codebase has a lot more levels of abstraction than other parts of sympy that I've looked at. I am still game to take a crack at this, but if you have suggestions e.g. "the first step is to create a subclass of X" that would be helpful.

**Comment by sylee957** [comment]
> This is obviously worse than my representation if we expect there to be high powers with many consecutive instances of the same generator

I think that this needs evidence. 
I also question a lot why Python community doesn’t use list or string like your idea, than not, and we are trying to implement the same (free) monoid just with different implementation, slightly more complicated, and may not be worth if they turn out to be much slower

**Comment by oscarbenjamin** [comment]
There is obviously a tradeoff between dense and sparse representations like ` x**3 y x**2 z` as
```
[(0, 3), (1, 1), (0, 2), (2, 1)]
```
or
```
[0, 0, 0, 1, 0, 0, 2]
```
Typically if you were writing this code in something like C then the second approach would be very compact and for simple cases (small exponents) would likely be faster than the list of tuples sort of representation. Typically in Python though the benefit of compactness in representation is reduced because even the flat list is actually an array of pointers to heap objects.

On the other hand it is not hard to go to the extreme where the flat list would be very inefficient for large exponents. I would say that unless there seems to be some significant speedup in common cases for the flat list approach then it is better to use the list of tuples approach.

> if you have suggestions e.g. "the first step is to create a subclass of X" that would be helpful

I would say don't subclass anything. If we want to reuse code then we should do that with functions rather than subclasses (two classes can call the same functions though). Using subclasses for code sharing generally leads to problems down the line.

First step is just to make a class with a few constructors, some domains and some arithmetic methods. Poly options like `field=True`, `extension=True`, `domain=QQ` etc to control the ground domain should work (these can be passed to `construct_domain`). It should also be possible to specify the generators.

Is there any concept of monomial ordering for noncommutative polynomials?

**Comment by sylee957** [comment]
I do think that there is easy way to 

- Polynomial, (or Commutative Sparse Polynomial as I fully qualify in this case) is just a dictionary of list of nonnegative integers to DomainElement.
  - In PEP484 typing notation, it can be precisely understood as `dict[list[int], DomainElement]`
- Now, we observe that `list[int]` is the self-contained representation of monomial, why not call this `Monomial` type
- Now, polynomial is just a type `dict[Monomial, DomainElement]`, I can also call it `dict[Monommial, Coefficient]`
- Now, we also observe that for commutative polynomial, `Monomial` is not only `list[int]` but also can be any commutative monoid. So commutative polynomial is just a dictionary from commutative monoid to coefficient.
  - [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) is another good representation for commutative monoid.
  - Diofant for example, uses similar definition as `Counter`, (see https://github.com/diofant/diofant/blob/master/diofant/polys/monomials.py). We should investigate if this is feasible, or regresses something about performance. However, I like the idea that diofant, starts to free the definition of monomial and polynomial, which gives the insight and allows generalization as below:
- Noncommutative polynomial, is simply a dictionary from (general) monoid to coefficient.
  - The example of general monoid is string, or list. In fact, string or list are known as 'free monoid' for some technical users, even though many mundane developers use it and understand practice algebra every day.
  - I also want to share the observation that `[(0, 3), (1, 1), (0, 2), (2, 1)]`, or whatever you call, is just an another representation of a list or string. as I have understood and questioned in https://github.com/sympy/sympy/issues/26114#issuecomment-1907821539.
  - So I also find that it can be nice if anyone implementing that, can just make its interface similar as list or string.
    For example, I observe that [`split`](https://github.com/sympy/sympy/pull/26093/files#diff-c81546f71d8aca2d6e6917530a56499438c683c010864b7ae5e3bed426385617) just resembles list slicing, and it could be more easier if anyone can implement that more 'catchy' way like list or string.
  - However, I do dislike some designs like if `Monomial` should contain coefficient. Not only because of the confusion with glossary in SymPy, (The definition of monomial in SymPy is something like $x^2y$) but also because it can make some things technically complicated, without any gain of generality. For example, I observe that [moving coefficient left and right](https://github.com/sympy/sympy/pull/26093/files#diff-c81546f71d8aca2d6e6917530a56499438c683c010864b7ae5e3bed426385617R1446-R1447) can be ambiguous and can just add technical complexity and does not add any value.
- So now we arrive to the discussion to how to implement negative exponents. Obviously, I would like to avoid inventing things like `factor` there because I'm not math genius. But obviously, we can make easy extensions like below (that just implement really basic algebra like add/mul like)
- For commutative polynomials, if we replace 'commutative monoid' by 'commutative group' for `Monomial`, then we get a polynomial of negative exponents
- For noncommutative polynomials, if we replace `monoid` by `(free) group`, then we get a noncommutative polynomial of negative exponents.
- Later, I may open a room to generalize `Monomial` to something as "fintely presented group/monoid modulo terminating and confluent writing system", which may look like some generalized abstract nonsense. However, I do think that this is useful because it opens up the room to implement something like 'a polynomimal where monomials have some nontrivial algebraic relationship between them', without much pain.

So in conclusion, what I suggest (in the most obvious direction) is that we should just extend existing sparse polynomial, and we just implement different type of `Monomial` (as described above, free monoid, free group, free commutative monoid, free commutative group) to fill in the different needs.

However, I do think that we have some things to concern about the project:

- First of all, we do have. For example, we not only deal with pure python implementation, but also need to deal with Flint wrapper implementation of polynomial. In flint wrapper, the coefficient and monomial are fused in the definition, and we had used to implement that as adhoc implementation of `dict[Monomial, ZZ]` or `dict[Monomial, QQ]`. And flint, may not even have anything like noncommutative polynomial. 
- Second of all, polys could be too 'fit' for commutative polynomials. A lot of its necessarily requirements (not only for `MonomialOps`, but also filling other interfaces like gcd, groebner basis, algebraic extensions, or even galois group) may be too fit for commutative polynomials, and can be much difficult for someone who implements this. So although my blueprint can be simple and easy, implementing this in SymPy mot be easy. Unfortunately, even though I have drawn a picture like this multiple times before, I did become a champion to implement this in SymPy because I know this can lead to lots of failures, and waste a lot of time.
  (Starting another toy project that specifically deals with this study, outside SymPy can be much more productive)
- I share the (opinionated) view that inheritance is generally bad. I do think that it's better to avoid inheritance at all here just because I or SymPy community, do not understand inheritance very well, and obviously had experienced a lot of problems of 'monkeys' abusing using inheritance. And I am not self-acclaimed guru who understands inheritance very well, sorry
  - The other observation I would share is that, if inheritance makes sense, commutative polynomial should inherit from noncommutative polynomial. And if you think in other way that noncommutative polynomial should inherit from the commutative polynomials, just to reuse some implementation, it may be wrong.

**Comment by zhempstead** [comment]
> However, I do dislike some designs like if Monomial should contain coefficient. Not only because of the confusion with glossary in SymPy, (The definition of monomial in SymPy is something like $x^2y$) but also because it can make some things technically complicated, without any gain of generality. For example, I observe that [moving coefficient left and right](https://github.com/sympy/sympy/pull/26093/files#diff-c81546f71d8aca2d6e6917530a56499438c683c010864b7ae5e3bed426385617R1446-R1447) can be ambiguous and can just add technical complexity and does not add any value.

Yes, when I wrote that I didn't realize that Sympy was consistent about monomials not having coefficients. If I were still working on that PR I would rename `Monomial` to `Term`. However, it's still necessary for the factoring algorithm to have a function that splits terms into two terms of the given degree, which requires making some (arbitrary) choice about the coefficient.

**Comment by sylee957** [comment]
If you think so, I can’t help much of the design further, unfortunately

I would assume that it does not ‘unify’ with existing implementation of sparse polynomial, 
It is implemented as a dictionary from monomial to coefficient, not as a list of ‘term’ (fused coefficient and monomial)

**Comment by zhempstead** [comment]
1. I'm not planning on implementing factoring as part of the initial suite of capabilities of this class - it will probably at first be limited to basic operations and really simple functions like `degree`.

2. When I say that the algorithm requires splitting terms into two, this doesn't lock in any particular implementation. For instance, we could have an NCPolynomial class which has a function `_split` that returns two NCPolynomials corresponding to splitting the leading term of the original `NCPolynomial`

**Comment by sylee957** [comment]
> When I say that the algorithm requires splitting terms into two, this doesn't restrict us to any particular implementation. For instance, we could have an NCPolynomial class which has a function _split that returns two NCPolynomials corresponding to splitting the leading term of the original NCPolynomial

What do you mean by this specifically? It sounds like factorization for me, if not precise, what else does spliting one polynomial to two polynomial corresponds to?

**Comment by oscarbenjamin** [comment]
> 1. really simple functions like `degree`

Is `degree` simple in the noncommutative case? Or are there potentially different ways of defining it that possibly depend on other things like choice of term order etc?

The concept of degree is already ambiguous in the multivariate commutative case so I would be surprised if it is simple the noncommutative case.

**Comment by zhempstead** [comment]
> What do you mean by this specifically? It sounds like factorization for me, if not precise, what else does spliting one polynomial to two polynomial corresponds to?

https://arxiv.org/pdf/1002.3180.pdf
Algorithm 1, line 1. In this paper, `monomial` includes the term. Basically I need a function or method that takes in a term and returns two terms of desired degree that, when multiplied together, equal the original term.

> Is degree simple in the noncommutative case? Or are there potentially different ways of defining it that possibly depend on other things like choice of term order etc?

Looking more carefully at Sympy's documentation, `total_degree` is the function that I meant. Surely this is pretty simple, right? Just the maximum monomial degree, and monomial degree is simple.

I agree that it's not clear whether the `degree` function would be meaningful for the noncommutative case, but why is it ambiguous in the commutative case? Isn't it just the maximum of the degree of each monomial, where the degree of the monomial is just the power of the specified generator.

**Comment by zhempstead** [comment]
@oscarbenjamin I realize I never answered this earlier question:
> Is there any concept of monomial ordering for noncommutative polynomials?
Yes, all the standard orderings work for monoids too. If anything, it's simpler since if we think of generators as letters, monoids are just words.

Although grevlex is simple to define, I wasn't sure if it was meaningful, but https://arxiv.org/pdf/1303.0920.pdf uses it with free algebras.

**Comment by oscarbenjamin** [comment]
The reason for my question about monomial ordering is just to consider whether it is something that should be baked in from the start.

SymPy currently has two representations for multivariate polynomials: DMP and sparse. The DMP representation is recursive like `R[x][y][...]` and so the structure implicitly uses lex order:
```
In [3]: Poly(x**2*y + x + y).rep
Out[3]: DMP_Python([[mpz(1), mpz(0)], [mpz(1)], [mpz(1), mpz(0)]], ZZ)
```
The sparse representation allows the order to be defined as part of the ring:
```
In [10]: R, x, y = ring("x, y", domain=QQ, order='lex')

In [11]: x**2*y + y**3*x + y
Out[11]: x**2*y + x*y**3 + y

In [12]: R, x, y = ring("x, y", domain=QQ, order='grevlex')

In [13]: x**2*y + y**3*x + y
Out[13]: x*y**3 + x**2*y + y
```
The order is baked in at a low level and affects many operations because it defines the behaviour of leading term, leading monomial etc:
```
In [16]: p = x**2*y + y**3*x + y

In [17]: p.LC
Out[17]: mpq(1,1)

In [18]: p.LT
Out[18]: ((1, 3), mpq(1,1))

In [19]: p.LM
Out[19]: (1, 3)
```
Other operations like Groebner basis etc are then determined by this monomial ordering as well as printing.

**Comment by oscarbenjamin** [comment]
> I agree that it's not clear whether the `degree` function would be meaningful for the noncommutative case, but why is it ambiguous in the commutative case? Isn't it just the maximum of the degree of each monomial, where the degree of the monomial is just the power of the specified generator?

There are different ways:
```
In [1]: p = Poly(x**5+y**7)

In [2]: p
Out[2]: Poly(x**5 + y**7, x, y, domain='ZZ')

In [3]: p.degree()
Out[3]: 5

In [4]: p.degree_list()
Out[4]: (5, 7)

In [5]: p.total_degree()
Out[5]: 7
```
Which of these is most meaningful depends on the context.

**Comment by zhempstead** [comment]
Sure, I can accept that it's ambiguous as to which one of these notions of `degree` is necessary in the given context, but each function seems pretty straightforward to implement given the precise definition.

**Comment by zhempstead** [comment]
> The reason for my question about monomial ordering is just to consider whether it is something that should be baked in from the start.

Thanks for the explanation about DMP. One big difference between polynomial rings and free algebras is that in the latter, we don't have as much flexibility to change the treatment of a variable, i.e. whether the variable is a generator or part of the ground domain.

Like if we have polynomial `x**2 + x*y` we can go back and forth between considering this a polynomial with generator `x` and monomials `x**2` and `x`, or a polynomial with generators `x` and `y` and monomials `x**2` and `x*y`.

We can't do this with elements of a free algebra, since the underlying ground domain is still commutative but the generators no longer are. So if `y` is commutative, it *must* be an element of the ground domain, and if `y` is non-commutative, it must be a generator.

---

I propose the following roadmap for the changes:
1. Implement a FreeAlgebraElement class whose interface is as similar to `Poly` as makes sense, implementing simple methods. Code will be reused as much as possible via functions.
2. Add a new PolyLike class which simply adds a constructor that will return either a `Poly` or a `FreeAlgebraElement` depending on the input arguments.
3. For functions like `factor` and `groebner` which would be meaningful for FreeAlgebra instances but the functions currently coerce their inputs into `Poly`:
  a. Implement a version of the function that works for `FreeAlgebraElement`
  b. Instead of having the main function construct a `Poly` instance, call the `PolyLike` constructor and then call the appropriate sub-function depending on the class of the resulting instance
  
 It *might* also be useful to add a new `is_Polylike` property - I'm not sure yet.

**Comment by oscarbenjamin** [comment]
> it's ambiguous as to which one of these notions of `degree` is necessary in the given context, but each function seems pretty straightforward to implement given the precise definition.

That is true but when designing an interface it is important to be aware of what the different possible meanings are before adding a function like "degree" and giving it a particular meaning. In so far as it makes sense it would be good if the interface for noncommutative polynomials matches the interface for commutative polynomials but obviously there will have to be some differences. Besides `total_degree` it is not clear to me if the other meanings carry over to the noncommutative case.

**Comment by oscarbenjamin** [comment]
The roadmap sounds reasonable. I'm not sure yet if it makes sense to transparently create noncommutative polynomials in things like `Poly(expr)`, `expr.as_poly()` etc. Ideally those would handle the commutative case automatically but there is probably a lot of code that would break if noncommuyative polynomials started appearing. Starting from arbitrary expressions we would probably often have situations in which there is a mix of commutative and noncommutative generators.

For functions like `as_poly` etc there are other reasons why we might want to redesign the exact interface that is usually used in SymPy besides just handling noncommutative polynomials.

First step I think is just to add a NCPoly class and then have `factor_nc` use it. Then it is only being used when explicitly requested.

**Comment by zhempstead** [comment]
> That is true but when designing an interface it is important to be aware of what the different possible meanings are before adding a function like "degree" and giving it a particular meaning. In so far as it makes sense it would be good if the interface for noncommutative polynomials matches the interface for commutative polynomials but obviously there will have to be some differences. Besides `total_degree` it is not clear to me if the other meanings carry over to the noncommutative case.

I agree. I think that we will adhere to the Poly interface in the sense that even if what Sympy calls `total_degree` is the only meaningful "degree" for an NCPoly, then we will still call the function `total_degree` to match `Poly`'s interface and simply raise an error if `degree` is called.

> I'm not sure yet if it makes sense to transparently create noncommutative polynomials in things like Poly(expr), expr.as_poly() etc.

Oh, I think it does *not* make sense to do this. A free algebra is not a polynomial ring, so if the user calls `Poly` they should never end up with a `FreeAlgebraElement`. Instead, if `Poly` detects a noncommutative term we should just update the error message to suggest that the user call the new `FAElement` or `expr.as_fa_element()` (or whatever the name is).

On the other hand, there are functions like `factor` and `groebner` that will, behind the scenes, attempt to convert an expression into a `Poly`, do the requisite calculation, and then convert back into an expression. *These* functions should be updated so that their behind-the-scenes conversion will possibly create a `FreeAlgebraElement` instead of a `Poly`, depending on the input.

(BTW at this point, after reading more on the topic, I am against the term `NCPoly` because while a free algebra is a non-commutative *analogue* of a polynomial ring, it is definitely *not* a polynomial ring, so `NCPoly` is a misnomer. I think the name of the class should reflect the proper mathematical name, i.e. "free algebra element." Beyond that I have no opinion on the name).

**Comment by zhempstead** [comment]
One thing I'm confused by is the existence of both `Poly` and `PolyElement`. The former uses a dense representation and the latter uses a sparse representation, but what is the purpose of having both as part of the API?

The latter seems more grounded in the `Domains` code, but the former subclasses `Basic` and thus has e.g. `is_Poly` as a field.

On top of this, `IPolys` in `sympy/poly/compatibility.py` seems to be a class that bridges both? But it's hardly used.

**Comment by zhempstead** [comment]
It seems like `Poly` will do its best to come up with an appropriate domain such that the input expression can "count" as a polynomial, whereas `PolyElement` requires one to first specify a domain and then use that domain to unambiguously construct an element from the input expression (or fail if the input expression is not valid for that domain).

I can see why both approaches are useful in different contexts, but what I don't understand is why `Poly` isn't just a function `poly` that ultimately returns an instance of `PolyElement`.

**Comment by oscarbenjamin** [comment]
Poly and its DMP representation were made first and PolyElement later. Some things have been changed to use PolyElement but some things have not. Some operations are not implemented directly for PolyElement and so PolyElement often converts to DMP to perform an operation.

In some situations the sparse representation provided by PolyElement is significantly more efficient e.g. for highly multivariate polynomial rings like `R[x,y,z,...]`. However for simple univariate or perhaps bivariate polynomials the semi-dense recursive representation provided by `DMP` can often be more efficient. It is not a trivial choice to choose between these two and many operations will convert between the representations either because some algorithm has not been implemented for one or the other or because it is more efficient in one representation rather than the other.

You can also choose between using PolyElement or DMP at the domain level like so:
```
In [5]: ZZ.old_poly_ring(x).one
Out[5]: DMP_Python([mpz(1)], ZZ)

In [6]: dict(ZZ.poly_ring(x).one)
Out[6]: {(0,): mpz(1)}
```
From a "user" perspective it is not common to use anything other than Poly although Poly itself will use PolyElement recursively when the ground domain is a polynomial because at the domain level PolyElement is always used:
```
In [11]: Poly(x**2*y + x, x)
Out[11]: Poly(y*x**2 + x, x, domain='ZZ[y]')

In [12]: Poly(x**2*y + x, x).rep
Out[12]: DMP_Python([y, 1, 0], ZZ[y])

In [13]: Poly(x**2*y + x, x).rep.to_list()
Out[13]: [y, 1, 0]

In [14]: dict(Poly(x**2*y + x, x).rep.to_list()[0])
Out[14]: {(1,): mpz(1)}
```
Currently Poly uses DMP as its internal representation and if the polynomial is univariate and the domain is ZZ, QQ or GF(p) then the DUP_Flint representation is used which is a wrapper around python-flint's `fmpz_poly`, `fmpq_poly`, `nmod_poly` and `fmpz_mod_poly` types. In the future when python-flint exposes Flint's multivariate polynomial types they will be used by Poly for multivariate polynomials and also in place of PolyElement at the domain level. The simple way to do this would be to make two separate wrapper classes but it would be nicer if they could be unified somehow so that `DMP` and `PolyElement` have the same interface. I am not sure if making them have the same interface is reasonable though and I actually wonder if univariate and multivariate polynomials should have different interfaces.
